1Password/extension-messaging
# internal(ex-msg): Add helper functions to improve developer experience and add documentation

## 🦾 Reason to Be

Prior to release we wanted to add some helper functions to make the developer experience better, and add support for encrypted values so users can send non-string values across to be saved.

## ✅ MR Checklist

- [x] **Properly Titled** The title of this MR is of the form `type(scope): Message goes here`. For example, `fixed(brain): We now fill correctly on google.com`. [More info in the project README](https://gitlab.1password.io/dev/core/core#changelog-entries).
- [ ] **Tests** You’ve written tests that prove your changes work correctly and will catch bugs introduced by future generations.
- [x] **Documentation** If you’re adding or improving a new internal API you’ve properly documented its public-facing components with good example usage.
- [ ] **Follow-on Issues** If this MR is providing stubs that other developers will need to implement, you’ve filed requisite issues for them.
- [ ] **Security Review** Asked for security review if this touches on cryptography, authentication, authorization, changes to logging, input validation or anything else that might impact security?

## 📝 Logging

N/A

## 🤔 Thought Process

- Add English strings for item templates instead of end user needing to provide template number
- Provide helper function to send requests to a specific extension ID
- Provide helper function to encrypt individual values in a save request
- Add documentation with unencrypted and encrypted requests, supported item templates, way for developers to contact us to get their extension added to the allowlist

**TODO:**
- Test save requests for each of our item templates to establish the scope of which item templates we are currently able to support
- Block item templates we do not currently support (ie `identity` for example still needs some work)
- Timeout + error handling
 
## 📋 How To Test

1. Build the extension demo app, follow steps [here](https://www.notion.so/1password/extension-messaging-932b9b9912464aaaaa1daef218ae8cf4#3e6a215767ca44988b34586103fb5930)
2. Checkout core branch `nima/implement-encrypted-value-type` and build b5x (remember to add the demo extension app ID to the list of allowed extensions as detailed in Notion)
3. Load the background app as an extension in Chrome and open the background
4. In the background console, run `window.sendCreateCryptoItem("your-local-b5x-extension-id-here")` > expected behavior is that a new tab should open in Chrome with a Save Dialog and a crypto wallet item with an encrypted recovery seed
5. Next, try following the steps in the README to create and save your own encrypted and unencrypted items in `demo-extension/src/background.ts`. Remember to build again after modifying this file.


## 🐛 Possible Impacts

In the future we may want to add support for fully encrypted requests as well as value-only like we're doing in this MR.
